### DISCUSSION TITLE
Problems using py5 in OSX

### DISCUSSION THREAD

**Discussion by Maltiec** [py5coding/py5generator#262] [Q&A] [answered]
Hello! Some of my students on OS X faced problems using py5. Among usual things like "can't find JAVA 17", we've got some strange error from PIL library (after starting py5 sketch): can't find @rpath/libtiff.5.dylib  We've located the missing file (it is in totally different path), but I don't know to add correct path to this variable (or may be there is some other way to solve the problem). py5 was installed with anaconda and  .yml file.
![photo_2023-03-11_12-18-29](https://user-images.githubusercontent.com/35667683/224547486-7079c250-11a2-444c-a76d-0f0870ac77e2.jpg)

**Comment by Maltiec** [comment, marked as answer]
UPD. Problem solved by copying libtiff file right into py5 directory

**Reply by hx2A** [reply]
Hello, @Maltiec !

That's an unusual PIL error. Are you using Anaconda environments for the install? That also might solve it.

Regarding Java version problems, I recommend debugging with `py5_tools.get_jvm_debug_info()` to see what jpype and py5 see. If Java 17 is installed but it can't be found, you can always set the `JAVA_HOME` environment variable.

**Reply by Maltiec** [reply]
yep that was installation in separate anaconda environment

**Reply by hx2A** [reply]
Interesting. Anaconda is supposed to provide those kinds of library files. The issue is probably with Anaconda & PIL. Glad you have it working though.

**Reply by Maltiec** [reply]
Ok, we've got the problem on M1, can't start jvm. I saw the issue with jvm on M1 and inctructed student to install jvm18 instead of 17, but still no luck

**Comment by hx2A** [comment]
The latest version of py5 should work fine on the M1 machines. Can you
confirm the right version is installed? Can you also try installing
Processing and seeing if that works?

**Reply by Maltiec** [reply]
0.8.3a1

**Reply by Maltiec** [reply]
vanilla Processing works fine

**Reply by hx2A** [reply]
Can I have more information about the error messages and symptoms?

What is the output of `py5_tools.get_jvm_debug_info()`?

Can you tell me if this code works when run in the py5 environment?

```python
import jpype
jpype.startJVM()
```

**Reply by Maltiec** [reply]
We'll try that in a day or two and I send back the result.

**Reply by Maltiec** [reply]
ok, I think I know what is the problem. This code:
import jpype
jpype.startJVM()
resulted in:
Traceback (most recent call last):

  File ~/opt/anaconda3/envs/py5coding/lib/python3.8/site-packages/spyder_kernels/py3compat.py:356 in compat_exec
    exec(code, globals, locals)

  File ~/.spyder-py3/temp.py:5
    jpype.startJVM()

  File ~/opt/anaconda3/envs/py5coding/lib/python3.8/site-packages/jpype/_core.py:224 in startJVM
    _jpype.startup(jvmpath, tuple(args),

OSError: [Errno 0] JVM DLL not found: /Library/Java/JavaVirtualMachines/jdk-18.0.2.1.jdk/Contents/Home/lib/libjli.dylib

then py5_tools.get_jvm_debug_ino() resulted in:
default jvm path: /Library/Java/JavaVirtualMachines/jdk-18.0.2.1.jdk/Contents/Home/lib/libjli.dylib
JAVA_HOME environment variable  <not set>
jvm version (0,0,0)

We've checked the default path for java and there is no /Library dir, but /Libraries! I suppose we need to make an environment variable JAVA_HOME. What is the shortest way to make it in OS X?

**Reply by hx2A** [reply]
On my Linux machine here I have this in my `~/.bashrc` file:

```bash
export JAVA_HOME=/usr/lib/jvm/java-17-openjdk
```

OSX uses something else besides bash (zsh?) and has its own set of configuration files. I would poke around in the home directory to find config files (with `ls -a`) and find a good spot for it, perhaps in one that has other environment variables set. Remember to start a new terminal after making a change to those files.

**Reply by Maltiec** [reply]
yep it is so simple on our Linux machines ) OS X is a somekind of corrupted Linux to me.)

**Reply by hx2A** [reply]
> yep it is so simple on our Linux machines ) OS X is a somekind of corrupted Linux to me.)

Haha, indeed it is!

I just checked my OSX machine and I see I set `JAVA_HOME` in `~/.zshrc`.

**Reply by Maltiec** [reply]
well, the problem is still there, it seems like jpype cannot see files in the directory, I've tried different variants, even installed "special for M1" JAVA - still no success. Starting using jpype telling me this:
OSError                                   Traceback (most recent call last)
Cell In[2], line 3
      1 get_ipython().run_line_magic('gui', 'osx')
      2 import jpype
----> 3 jpype.startJVM()

File ~/opt/anaconda3/envs/py5coding/lib/python3.8/site-packages/jpype/_core.py:224, in startJVM(*args, **kwargs)
    222 prior = [locale.getlocale(i) for i in categories]
    223 # Start the JVM
--> 224 _jpype.startup(jvmpath, tuple(args),
    225                ignoreUnrecognized, convertStrings, interrupt)
    226 # Collect required resources for operation
    227 initializeResources()

OSError: [Errno 0] JVM DLL not found: /Users/bozigna/Documents/Java/JavaVirtualMachines/temurin-17.jdk/Contents/Home/lib/libjli.dylib

**Reply by Maltiec** [reply]
I've searched on similar issues and found this: https://github.com/jpype-project/jpype/issues/994

**Reply by hx2A** [reply]
I'm not familiar with the temurin JDK. Is the file libjli.dylib anywhere to be found in the install directory?

Can you try installing OpenJDK, setting `JAVA_HOME` to point to that, and then testing this again?

There are some JDKs that do not work with jpype:

https://jpype.readthedocs.io/en/latest/userguide.html#unsupported-java-virtual-machines

temurin is not mentioned, but I know that OpenJDK works, so give that a try.

You can also let install-jdk do the install, as mentioned in the py5 install instructions.

```bash
pip install install-jdk
python -c "import jdk; print('Java installed to', jdk.install('17'))"
```

For this approach, you don't need to set `JAVA_HOME`, and if you don't have it set, it will pick the install-jdk JDK over others installed elsewhere on the machine.

**Reply by Maltiec** [reply]
ok, we'll try this path today.  libjli.dylib - is in the Contents/Home/lib I've checked it

**Reply by Maltiec** [reply]
we finally did it! the only problem is that import jdk raises ["TypeError: 'type' object is not subscriptable"], this is due python in dependecies for py5 is 3.8, I've read that it is solved in Python > 3.9, so I installed Python 3.9.16, then jdk, then added JAVA_HOME variable. And it works!

**Reply by hx2A** [reply]
> And it works!

Excellent! The install-jdk library is reliable and py5 is designed to look for install-jdk installations. It is a good avenue for addressing computers with jdk install problems.

> the only problem is that import jdk raises ["TypeError: 'type' object is not subscriptable"]

It is unclear if install-jdk should support 3.8 or not. I just opened an [issue](https://github.com/jyksnw/install-jdk/issues/83) with that project to find out.
